0004-market-integration-token-standard.md

```diff
@@ -0,0 +1,123 @@
+- Proposal Name: "marketplace_integrated_token_standard"
+- Start Date: 2019-06-24
+- NEP PR: [nearprotocol/neps#0000](https://github.com/nearprotocol/neps/pull/0000)
+
+# Summary
+[summary]: #summary
+
+A standard interface for non-fungible tokens allowing for ownership and transfer, specifically targeting third-party marketplace integration.
+
+# Motivation
+[motivation]: #motivation
+
+Non-fungible tokens (NFTs) have been described in many ways: digital goods, collectible items, unique online assets etc. The current core use case for NFTs is speculating on value and trading unique items. The use case of trading NFTs should natively supported for this emerging market. 
+
+The idea of a marketplace-integrated NFT is to make it as easy as possible to merge new services with existing tokens that use the standard. Existing NFT standards still require some customization on the part of third party integrators, which becomes a burden with each NFT integrated.  This standard will allow anyone to build a marketplace on top of any number of tokens, not just allowing but promoting trading and extension. These marketplaces can be built into games and apps that use the specific metadata in whatever way they choose.
+
+Prior art:
+ERC-20 standard: https://eips.ethereum.org/EIPS/eip-20
+ERC-721 standard: https://eips.ethereum.org/EIPS/eip-721
+ERC-1155 standard: https://eips.ethereum.org/EIPS/eip-1155
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Example of minting a token
+
+```TypeScript
+[...]
+let corgiToken = generateRandomCorgiToken(tokenDetails);
+// corgiToken.id === null 
+let mintedCorgiToken = mintToken(corgiToken);
+near.log(mintedCorgiToken.tokenType.name);
+// "CorgiToken"
+
+getTokenValue(corgiToken.id)
+// In most cases, value should be determined by market, however there are cases where it is important that value be set by the developer.
+
+```
+  
+For user-facing NEPs this section should focus on user stories.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The NEAR marketplace-integrated token standard
+
+Template for smart contract.
+Note that as of this writing, interfaces are unimplemented in AssemblyScript.
+
+```TypeScript
+class MarketStandardToken {
+  getTokenTypes():TokenType[]
+  getToken(tokenId:string):Token
+  getTokensByOwner(ownerId:string):Token[]
+
+  // getCount is equivalent of balanceOf
+  getCount(tokenType: TokenType, ownerId: string)
+  getTokenValue(tokenId:string):u64
+  getTokenData(tokenId:string):bytes
+
+  // It's up to the developer of a token to decide how tokens are generated.
+  // Minting should return the Token with a unique ID attached.
+  mintToken(token:Token):Token
+
+  // Note that transferring can be done on 1 or more tokens
+  transfer(recipientId:string, ownerId:string, tokenIds:string[]):void
```

Comment:
When we discussed this for ERC-20 example, we realized we need a callback to the receiver's contract. E.g.
On ERC-20 you would have a function:

// Token's owner would be the immediate sender_id (that initiated this call).

    transfer(receiver_id: string, amount: u128, method_name: string, args: bytes): void

The contract should initiate an async call to `receiver` with `method_name`, and pass the `sender_id`, `amount` sent and `args`. And revert this transfer if the callback fails with error. This ensures composability of the contracts in the async settings. While the transfer is in progress, the amount is locked and can't be withdrawn.

Reply:
Also we may want to separate `new_owner_id` and the `callback_receiver_id` in the transfer method.
Maybe we need to have 2 functions, simple transfer and transfer with callback